### PR TITLE
Replica byte lag health check

### PR DESCRIPTION
**Pull request by benschinn**
Addresses issue https://github.com/film42/pgreba/issues/3.

Reads `pg_current_wal_lsn` from primary and `pg_last_wal_receive_lsn` from replica and gets byte lag from ` pg_wal_lsn_diff(lsn, lsn)`. 

```
postgres=# select pg_current_wal_lsn();                                                                                                                 
-[ RECORD 1 ]------+----------                                                                                                                          
pg_current_wal_lsn | 0/4000108 
```

```
postgres=# select pg_last_wal_receive_lsn();                                                                                                            
-[ RECORD 1 ]-----------+----------                                                                                                                     
pg_last_wal_receive_lsn | 0/4000108  
```

```
postgres=# postgres=# select pg_wal_lsn_diff('0/4000108', '0/4000108');                                                                                 
-[ RECORD 1 ]---+--                                                                                                                                     
pg_wal_lsn_diff | 0  
```

example payload:
```
{
   "state":0,
   "postmaster_start_time":"2020-08-07 08:54:37.141 MDT",
   "role":"replica",
   "xlog":{
      "location":0,
      "received_location":1694501464,
      "replayed_location":1694501464,
      "replayed_timestamp":null,
      "paused":false
   },
   "replication":[

   ],
   "byte_lag: 0
}
```

### CODE REVIEW

File: data_source.go [hunk 1]
```diff
@@ -73,6 +91,7 @@ type NodeInfo struct {
 	Role                string             `json:"role"`
 	Xlog                *XlogInfo          `json:"xlog"`
 	Replication         []*ReplicationInfo `json:"replication"`
+	PgCurrentWalLsn     string             `json:"pg_current_wal_lsn"`
```
Comment (benschinn) [inline data_source.go]: Todo: remove this field out of NodeInfo struct. calculate the byte lag using `pg_wal_lsn_diff`.
https://severalnines.com/database-blog/what-look-if-your-postgresql-replication-lagging

File: config/config.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	}
+	c := new(Config)
+	err = yaml.Unmarshal(bytes, c)
+	if err != nil {
+		return nil, err
+	}
+	return c, err
+}
+
+func Conninfo(path string) (conninfo string) {
```
Comment (film42) [inline config/config.go]: Can we make this a method on `*Config`? 

Then we can pass this config struct around, which makes adding config bits much easier.

File: config.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+---
+host: localhost
```
Comment (film42) [inline config.yml]: Can we move this to an `examples` directory?

**Comment by film42** [comment]
Dude! Awesome work here! Left two comments, but this is really really good!

File: config.yml (hunk 3 above)
Comment (benschinn) [reply config.yml]: yes!

File: main.go [hunk 4]
```diff
@@ -77,7 +78,8 @@ func (hc *HealthCheckWebService) apiGetIsReplica(w http.ResponseWriter, r *http.
 }
 
 var (
-	defaultConnInfo = "host=localhost database=postgres user=postgres sslmode=disable binary_parameters=yes"
+	cfg, _err       = config.ParseConfig("./examples/config.yml")
```
Comment (benschinn) [inline main.go]: After adding `*Config` as the receiver to `Conninfo` func, I am initializing the config here. I am not doing anything with the error here. I found out that I can't use `if` construct outside a function body. Is it okay to leave it like this? Or should I create a function to handle the potential error value?

File: config/config.go (hunk 2 above)
Comment (benschinn) [reply config/config.go]: done.

File: main.go (hunk 4 above)
Comment (film42) [reply main.go]: Let's parse the config in the main func, that way we can pass that one instance of config around. For starters we probably want to make the pg replication data source take the config as an arg:

```
func main() {
  cfg, err := config.ParseConfig("./examples/config.yml")
  if err != nil {
    panic(err)
  }
  ds, err := NewPgReplicationDataSource(cfg)
}
```

That makes it easier to add additional args to these structs as time goes on. It also means you don't need to worry about error handling/ parsing this throughout the code.

Example:

```
func buildConninfo(conninfo map[string]string) string {
	cfg, err := config.ParseConfig("./examples/config.yml")
	if err != nil {
		log.Fatalln("Error parsing config:", err)
	}
	ci := fmt.Sprintf("host=%s port=%s database=%s user=%s sslmode=%s binary_parameters=%s", conninfo["host"], conninfo["port"], cfg.Database, cfg.User, cfg.Sslmode, cfg.BinaryParameters)
	return ci
}
```
can become:

```
func  (ds *pgDataSource) buildConninfo(conninfo map[string]string) string {
  return fmt.Sprintf("host=%s port=%s database=%s user=%s sslmode=%s binary_parameters=%s", conninfo["host"], conninfo["port"], 
    ds.cfg.Database, ds.cfg.User, ds.cfg.Sslmode, ds.cfg.BinaryParameters)
}
```

File: data_source.go [hunk 5]
```diff
@@ -193,9 +212,101 @@ FROM
...
+	if err != nil {
+		log.Fatalln("Error getting pg_wal_lsn_diff:", err)
+	}
+
+	nodeInfo.ByteLag = byteLag
+
 	return nodeInfo, nil
 }
 
+func (ds *pgDataSource) getConnInfo() (string, error) {
```
Comment (film42) [inline data_source.go]: Should we call this `getUpstreamConnInfo` or `getWalSenderConnInfo`? Something to indicate this conninfo is not ours, but the server sending us data.

Comment (benschinn) [reply data_source.go]: Great point. Renamed to `getUpstreamConnInfo`.

File: main.go (hunk 4 above)
Comment (benschinn) [reply main.go]: Awesome, I think that looks a lot better. Great advice. Let me know if it you have any other suggestions or nits!

**Review by film42 — APPROVED** [review]

**Comment by film42** [comment]
Looks great! Thanks for your work on this! Enjoy the PNW :)

**Comment by benschinn** [comment]
Thanks @film42 for reviewing and providing feedback!

**Pull request merged**